### PR TITLE
Add divexact!, lcm!, submul!

### PR DESCRIPTION
**Pull request by fingolfin**
Also refactor the code defining them to use macros to reduce
repetition substantially.

Finally merge and update the sections discussion unsafe operators
in the "rings" and "ring interface" chapters of the manual: the latter
now

### CODE REVIEW

File: src/exports.jl [hunk 1]
```diff
@@ -566,6 +568,7 @@ export weights
 export with_unicode
 export zero
 export zero!
+export submul!
```
Comment (lgoettgens) [inline src/exports.jl:571]: this should be moved to the alphabetical correct place

File: src/fundamental_interface.jl [hunk 2]
```diff
@@ -302,177 +302,96 @@ const VarName = Union{Symbol, AbstractString, Char}
 @doc raw"""
     zero!(a)
 
-Return the zero of `parent(a)`, possibly modifying the object `a` in the process.
+Return the `zero(parent(a))`, possibly modifying the object `a` in the process.
```
Comment (lgoettgens) [inline src/fundamental_interface.jl:305]:
```suggestion
Return `zero(parent(a))`, possibly modifying the object `a` in the process.
```

File: src/fundamental_interface.jl [hunk 3]
```diff
@@ -302,177 +302,96 @@ const VarName = Union{Symbol, AbstractString, Char}
...
-"""
-sub!(a, b) = sub!(a, a, b)
+zero!(a) = zero(parent(a))
 
 @doc raw"""
-    neg!(a, b)
+    one!(a)
 
-Return `-b`, possibly modifying the object `a` in the process.
+Return the `one(parent(a))`, possibly modifying the object `a` in the process.
```
Comment (lgoettgens) [inline src/fundamental_interface.jl:312]:
```suggestion
Return `one(parent(a))`, possibly modifying the object `a` in the process.
```

**Review by lgoettgens — APPROVED** [review]
apart from the comments, fine with me. I am not a big fan of `@eval` for these small loops, but we can of course unroll them again later, if we want to change only one of them

File: src/exports.jl (hunk 1 above)
Comment (fingolfin) [reply src/exports.jl:571]: ouch, how did that happen!?!? thank you

**Comment by fingolfin** [comment]
Comments addressed.

I don't love `@eval` for this either (it makes it harder to grep for declarations) I went for it here because I want to have the argument names consistent (`z,a,b`) and also the texts and there was a *ton* of repetition which is also error prone. In the end this saved me a lot of time and is more likely to be correct. If someone wants to unroll it again down the road, I don't mind, as long as I don't have to update the duplicated docstrings and implementations again later on :-).

The current setup also makes it super easy to add more of these... e.g. `remove!` and `valuation!` might be candidates. (Also `set!(a,b)` but we don't have a eval loop for that -- yet)

**Comment by lgoettgens** [comment]
Can you add HeckeCI, NemoCI and SingularCI to the "required" list? I am afraid to use auto-merge otherwise as it could just ignore those CI results.

**Pull request merged**